### PR TITLE
fix(TDI-43497): handle more lists than batch limit

### PR DESCRIPTION
**Pull request by undx**
https://jira.talendforge.org/browse/TDI-43497

### CODE REVIEW

**Comment by undx** [comment]
and a bit more shorter !!! Java naming syndrom :-)

File: marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java [hunk 1]
```diff
@@ -203,15 +207,25 @@ public SuggestionValues getActivities(@Option final MarketoDataStore dataStore)
     @Suggestions(LIST_NAMES)
     public SuggestionValues getListNames(@Option final MarketoDataStore dataStore) {
         log.debug("[getListNames] {}.", dataStore);
+        List<Item> lists = new ArrayList<>();
+        String nextPageToken = null;
+        Predicate<JsonObject> hasResults = j -> j.getJsonArray(ATTR_RESULT) != null;
+        Predicate<JsonObject> hasNextPageToken = j -> j.getString(ATTR_NEXT_PAGE_TOKEN, null) != null;
+        Consumer<JsonObject> listConsumer = l -> lists.add(new Item(String.valueOf(l.getInt(ATTR_ID)), l.getString(ATTR_NAME)));
```
Comment (clesaec) [inline marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java]: JsonObject.getInt(ATTR_ID) throw an Exception if not exists.
May be better to write a private function that transform JsonObject to Item ?

File: marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java [hunk 2]
```diff
@@ -203,15 +207,25 @@ public SuggestionValues getActivities(@Option final MarketoDataStore dataStore)
...
-                listNames.add(new SuggestionValues.Item(String.valueOf(l.getInt(ATTR_ID)), l.getString(ATTR_NAME)));
+            JsonObject result = handleResponse(listClient.getLists(aToken, nextPageToken, null, "", "", ""));
+            nextPageToken = result.getString(ATTR_NEXT_PAGE_TOKEN, null);
+            if (hasResults.test(result)) {
+                result.getJsonArray(ATTR_RESULT).getValuesAs(JsonObject.class).forEach(listConsumer);
+            }
+            while (hasNextPageToken.test(result)) {
+                result = handleResponse(listClient.getLists(aToken, nextPageToken, null, "", "", ""));
+                nextPageToken = result.getString(ATTR_NEXT_PAGE_TOKEN, null);
+                result.getJsonArray(ATTR_RESULT).getValuesAs(JsonObject.class).forEach(listConsumer);
```
Comment (clesaec) [inline marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java]: Miss "hasResults.test(result)"
think about create a private function to encaspulate repeat code

```            
      JsonObject result = handleResponse(listClient.getLists(aToken, nextPageToken, null, "", "", ""));
      nextPageToken = result.getString(ATTR_NEXT_PAGE_TOKEN, null);
      if (hasResults.test(result)) {
            return result.getJsonArray(ATTR_RESULT).getValuesAs(JsonObject.class);
      }
      return Collections.emptyList();
```
call it `getNextPage`

Comment (undx) [reply marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java]: if nextpagetoken is defined and there's not error in request (handleResponse), we don't need this predicate again. It's only needed for the first call which may be successfull but have no entries.

Comment (undx) [reply marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java]: Wanted to preserve the local scope of variables. we'll look to refactor anyway.

File: marketo/src/main/java/org/talend/components/marketo/input/LeadSource.java [hunk 3]
```diff
@@ -132,15 +132,16 @@ private JsonObject getLeadActivities() {
             activityTypeIds = configuration.getDataSet().getActivityTypeIds().stream().collect(joining(","));
         }
         String listId = configuration.getDataSet().getListId();
-        return handleResponse(leadClient.getLeadActivities(accessToken, nextPageToken, activityTypeIds, "", listId, ""));
+        return marketoService
+                .handleResponse(leadClient.getLeadActivities(accessToken, nextPageToken, activityTypeIds, "", listId, ""));
```
[the remark this answers is not in the document]
Comment (undx) [reply marketo/src/main/java/org/talend/components/marketo/input/LeadSource.java]: Will have a look but pagination is already handled in source comp.

File: marketo/src/main/java/org/talend/components/marketo/input/LeadSource.java [hunk 4]
```diff
@@ -72,7 +72,7 @@ public JsonObject runAction() {
     private JsonObject getLeadsByListId() {
         Integer listId = Integer.parseInt(configuration.getDataSet().getListId());
         String fields = schema.keySet().stream().collect(Collectors.joining(","));
-        return handleResponse(listClient.getLeadsByListId(accessToken, nextPageToken, listId, fields));
+        return marketoService.handleResponse(listClient.getLeadsByListId(accessToken, nextPageToken, listId, fields));
```
[the remark this answers is not in the document]
Comment (undx) [reply marketo/src/main/java/org/talend/components/marketo/input/LeadSource.java]: The scope of these generic handling functions have broaden to outside the scope of SourceOrSink.

File: marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java (hunk 1 above)
Comment (undx) [reply marketo/src/main/java/org/talend/components/marketo/service/UIActionService.java]: Actually, if the previous checks were successful, there's no reason that it do not exists. Mkto maintains integrity of its lists.

**Pull request merged**